File: atcoder/dp/G-Longest-Path.py
import sys
sys.setrecursionlimit(10**9)

n,m = map(int, input().split())
g = [[] for _ in range(n+1)]
for _ in range(m):
    node,neighbor = map(int, input().split())
    g[node].append(neighbor)

# dp[i] is longest path starting from node i which
# is the longest path starting from neighbor j + 1...
dp = [0] * (n+1)

def dfs(i):
    # we've already solved this node
    if dp[i] != 0:
        return dp[i]
    ret = 0
    for j in g[i]:
        ret = max(ret, dfs(j) + 1)
    dp[i] = ret
    return dp[i]
    
ans = 0
for i in range(1, n+1):
    ans = max(ans, dfs(i))
print(ans)

File: atcoder/dp/G-Longest-Path.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

vector<int> g[100000+5];
int dp[100000+5]; // dp[i] longest path from node i

int dfs(int i) {
    if (dp[i] > 0) return dp[i];
    // process node i
    for (auto j: g[i]) {
        dp[i] = max(dp[i], dfs(j) + 1);
    }
    return dp[i];
}

int main() {
    int n, m;
    cin >> n >> m;
    for (int i = 0; i < m; i++) {
        int node, neighbor;
        cin >> node >> neighbor;
        g[node].push_back(neighbor);
    }

    int ans = 0;
    for (int i = 1; i <= n; i++) {
        ans = max(ans, dfs(i));
    }
    cout << ans << endl;

    return 0;
}
